Query Loop: Add useBlockPreview, fix Query Loop wide alignment in the editor

andrewserong opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

This is an alternative to #35932 and is intended to resolve #33248.

It pulls in the `useBlockPreview` and `useDisabled` hooks from #35863 and uses them to render a block preview for the Post Template block, for those instances within the query loop that are not currently selected (and are therefore not interactive). By using these hooks, we can avoid adding the DOM nesting involved in using the `BlockPreview` component directly.

The `useDisabled` hook is mostly comprised of logic borrowed from the [Disabled](https://github.com/wordpress/gutenberg/blob/a7e2895829c16ecd77a5ba22d84f1dee1cfb0977/packages/components/src/disabled/index.js#L1) component. The context behaviour has not been included in the hook version, because we don't (yet) have a need for it. In a future follow-up, we could refactor the `Disabled` component to use the `useDisabled` hook introduced here, internally.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

In a page, insert a Query Loop block and set the alignment to Wide and make sure that the number of items within the loop is at least 2-3. Within the query loop block, select between the instances within the loop. Unlike in the reported issue (#33248), selecting between blocks should preserve the Wide layout.

__Note: this is specifically about dealing with adding a Query Loop block to a page — behaviour in the site editor should be unaffected__

## Screenshots <!-- if applicable -->

| Before | After |
| --- | --- |
| ![Kapture 2021-11-12 at 16 38 49](https://user-images.githubusercontent.com/14988353/141415852-af4e7eb8-c1c1-43c5-929a-cd11bb35e832.gif) | ![Kapture 2021-11-12 at 16 34 40](https://user-images.githubusercontent.com/14988353/141415466-94fe71d7-5bdc-42e4-853c-94ee486fa7d2.gif) |

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

andrewserong committed:
Add useBlockPreview, fix Query Loop wide alignment in the editor

andrewserong committed:
Add test for useDisabled

andrewserong committed:
Add minimal test for useBlockPreview hook

andrewserong committed:
Add useDisabled test to cover updates to a component

andrewserong committed:
Update useBlockPreview test to ensure elements within the block content are disabled

andrewserong committed:
Switch queryBy to getBy to ensure check for block is stricter

andrewserong committed:
Remove comment

andrewserong commented:
I think this is ready for review now, so I've switched it out of draft.

Mamaduka commented on packages/block-editor/src/components/index.js:
```diff
@@ -104,7 +104,7 @@ export { default as BlockList } from './block-list';
 export { LayoutStyle as __experimentalLayoutStyle } from './block-list/layout';
 export { default as BlockMover } from './block-mover';
-export { default as BlockPreview } from './block-preview';
+export { default as BlockPreview, useBlockPreview } from './block-preview';
```
Hi, @andrewserong. Should we export this new hook as experimental for now? Similar to `useDisabled`?

andrewserong replied:
Yes, of course, that's a good idea. Thanks George, I've updated it in https://github.com/WordPress/gutenberg/pull/36431/commits/7c8af494d3d0310b82a345be678645a2b5e1e3ff to export as `__experimentalUseBlockPreview` which also removes it from the readme, too.

Mamaduka replied:
Thanks, @andrewserong. I will try to do more detailed testing later today.

andrewserong committed:
Export useBlockPreview as __experimentalUseBlockPreview

andrewserong commented:
I've just given this a rebase to fix conflicts.

getsource commented:
Hi! Thanks so much for putting this together.

I did a bit of testing, and it does look to solve the issue!

I noticed a bit of changed behavior, though. Whether this is better or worse than the existing align jumping, I don't know -- but I figured it was worth documenting.

It's hard to describe, but there are a two things I noticed when selecting / unselecting different posts. I wasn't able to pinpoint the exact situations that cause it, but if I clicked in and out from them couple times, I could usually cause it to happen:
- There's bit of visual flicking / refreshing sometimes, where you can see the other items flash by (or load perhaps?). This comes with it feeling a little slower to respond
- Sometimes the view scrolls up to the top of the page or down, in situations where it would stay in the same place without the PR.

Before:

https://user-images.githubusercontent.com/1034160/143421685-c89816ea-aa87-40b9-b90b-ca7eaa792c43.mov

After:

https://user-images.githubusercontent.com/1034160/143421730-ff16edaf-aa90-4c83-872a-59eec66781a6.mov

Tested with Firefox 94.0.1 on OSX and Twenty-Twenty Two.

andrewserong commented:
Thanks so much for giving this a test @getsource and including the demo videos, that's super helpful! I don't think I tested this while using a background colour, so didn't notice the flickering issue (nor the scroll issue either 🤔). I'll investigate further (if I don't get time today, I'll follow up next week). Cheers!

andrewserong committed:
Optimise block preview via memoizing the component, and always rendering a preview for each block context

andrewserong commented:
It took me a little longer than expected to fix up the flickering issue. In https://github.com/WordPress/gutenberg/pull/36431/commits/8a34723d6436770dc39ac061018893d1f92e6a9b I've optimised the `PostTemplateBlockPreview` component's rendering by doing the following:

* Memoize the component
* Pass consistent values and references to the component (e.g. no anonymous function for `onClick`) to ensure the component doesn't re-render too frequently
* Always render a block preview for each block context in the list — but use `display: none` to hide the active preview. This ensures that when switching between block contexts, we switch visibility via CSS rather than requiring the preview itself to be re-rendered from scratch, which appears to have been the main culprit in the flickering issue that was most visible in Firefox (I had a harder time reproducing it in Chrome).

I believe this is now working pretty well, but could do with some more testing, as I _just_ got this finished before wrapping up my week. CC: @getsource.

| Before | After |
| --- | --- |
| ![Kapture 2021-12-03 at 17 26 11](https://user-images.githubusercontent.com/14988353/144555881-ee035754-0b4a-4723-ba0f-d027547b7710.gif) | ![Kapture 2021-12-03 at 17 24 39](https://user-images.githubusercontent.com/14988353/144555647-d5620de3-d135-400f-b7f1-417b7947fbc3.gif) |

ramonjd commented:
This is working as described for me I cannot detect any layout change with selecting inner blocks within the Query Loop block in the post editor.

![Kapture 2021-12-08 at 09 14 50](https://user-images.githubusercontent.com/6458278/145116504-691e8aba-7761-47b0-be3e-23b7fd684ace.gif)

It's not within the purview of this PR, but I'm not entirely sure that it addresses the Comments Query Loop block issue in https://github.com/WordPress/gutenberg/issues/37154

![Kapture 2021-12-08 at 09 34 36](https://user-images.githubusercontent.com/6458278/145116654-d4aac7ea-74bd-461d-ad3b-b722250ad69c.gif)

ramonjd commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -63,3 +69,58 @@ export function BlockPreview( {
+	);
+	const disabledRef = useDisabled();
+	const ref = useMergeRefs( [ props.ref, disabledRef ] );
+	const settings = useMemo( () => {
```
Feel free to ignore if you think it's more readable as is. 

```suggestion
	() => ( { ...originalSettings, __experimentalBlockPatterns: [] } ),
```

andrewserong replied:
That does look more readable to me — I think from memory I'd just copied this bit from the other BlockPreview component, so the shorter syntax looks nice. I'll follow up.

andrewserong replied:
Pushed in https://github.com/WordPress/gutenberg/pull/36431/commits/a00533ecad7f0721af1103d62a6deb48320d1e87 👍

ramonjd commented:
I can't comment on the implementation in great detail, but, having played with trunk, this is a clear improvement and I'm happy to approve once folks with more high-level context are happy with it.

andrewserong commented:
Thanks for testing @ramonjd!

> It's not within the purview of this PR, but I'm not entirely sure that it addresses the Comments Query Loop block issue in #37154

Yes, the Comments Query Loop block is quite similar so if this PR ends up being accepted, I think we'd then look at a follow-up for that block to deal with that issue 🤞.

andrewserong committed:
Tidy up settings useMemo

Co-authored-by: Ramon <ramonjd@users.noreply.github.com>

getsource commented:
Oh! I'm so sorry, I missed the mention for re-testing on this one. It looks like it's been tested now, but I'll take another look on Monday if it hasn't already been merged before then.

andrewserong commented:
> Oh! I'm so sorry, I missed the mention for re-testing on this one.

Thanks @getsource! No worries, I've had plenty of other things to keep me busy 😄 It'd be great to get another review/test to confirm the approach as both block previews and the Query Loop block can have quite a few edge cases. Also, since this uses a new hook-based approach for rendering the preview and the disabled behaviour, it'd be good make sure I haven't missed anything from the existing component-based approach, that might be needed here.

getsource reviewed (commented):
Okay! I went through and re-tested this, and the flickering seems to be resolved, along with the layout change being fixed.

This is great! In general, everything feels much smoother than pre-PR/trunk.

I don't think I'm the best person for a final code review, so submitting as Comment for that reason. Thanks so much!

gziolo reviewed (commented):
I would be in favor of giving it a try based on the feedback shared during testing. That's the most important aspect of this enhancement. We are planning to recreate a similar set of changes for the Comments Query Loop block as documented in https://github.com/WordPress/gutenberg/issues/37154. That should be a good way to validate whether the approach takes scales well.

This PR introduced a new experimental API in `@wordpress/components` – `useDisabled` hook. Is that a necessary step at the start or can it be inlined in the `@wordpress/block-editor` package for now? Do we have more use cases where the same hook could be used? @youknowriad and @ellatrix what do you think about it? If you think it's going to be useful in other places, we should include the documentation so the community knows how to use it.

andrewserong committed:
Add documentation for useDisabled hook

andrewserong commented:
Thanks for testing, folks!

> This PR introduced a new experimental API in `@wordpress/components` – `useDisabled` hook. Is that a necessary step at the start or can it be inlined in the `@wordpress/block-editor` package for now? Do we have more use cases where the same hook could be used?

Good question @gziolo! I originally added this hook [in a previous PR](https://github.com/WordPress/gutenberg/pull/35863) that didn't land, and over there @youknowriad [had the suggestion](https://github.com/WordPress/gutenberg/pull/35863#discussion_r738995769) of placing it in `@wordpress/compose`, which I've done here. The longer-term goal is to refactor the `Disabled` component in `@wordpress/components` to use this hook internally so that there's less duplication of logic (this hook borrows its logic from that component). I think the hook is generic enough that it could potentially be used in other circumstances, too, where a component might need to be rendered in a disabled state (disarming a form, perhaps?), so 🤞 the hook will sit nicely there alongside hooks like `useConstrainedTabbing`. Happy to move it if folks feel strongly about where it should go, though!

I like the suggestion of adding documentation, ~so I've had a go at adding in an entry for it in the `compose/README.md` file in https://github.com/WordPress/gutenberg/pull/36431/commits/57461d448d61c62b7ba8697605774cdd0de9527f — I know we often avoid adding documentation for experimental features, but I also see that the components package sometimes includes experimental features with a notice that it's experimental, so I've used that pattern here as well.~ Update: I've moved the documentation to the hook in https://github.com/WordPress/gutenberg/pull/36431/commits/ee8881f029940d66ebdead8a0ded7ef57f12fcfb — I just noticed that in CI it complains if we add it manually to the readme, I think the git commit hook must have failed for me locally 😄

Let me know if you'd like further changes!

andrewserong committed:
Move documentation to the hook

gziolo commented:
> The longer-term goal is to refactor the Disabled component in @wordpress/components to use this hook internally so that there's less duplication of logic (this hook borrows its logic from that component).

That and the suggestion from @youknowriad was the missing bit of information. I think it's a solid plan and with the JSDoc comment included for the new hook, it should be enough to expose the documentation once the API is promoted to stable. I hope you can do it in a follow-up that refactors the `Disabled` component.

gziolo reviewed (approved):
Let's get this in. I'm not the best person to approve, but you only live once and I trust all other reviewers 😄 

Thank you everyone for an excellent team work to help @andrewserong to take it to the finish line 🎉

andrewserong commented:
> I think it's a solid plan and with the JSDoc comment included for the new hook, it should be enough to expose the documentation once the API is promoted to stable. I hope you can do it in a follow-up that refactors the Disabled component.

Yes, of course! There's a few other tasks to wrap up first, but my goal is to follow-up with a refactor for the Disabled component in the new year. (In the shorter-term, I'd like to continue refining the click behaviour when switching between blocks and block previews in the instances of the loop). Thanks again for the review and merge!

ellatrix commented:
@andrewserong I don't get at all why we must render all inner blocks twice to fix and alignment issue. Could you try to explain this to me? I don't know anything about this block.

andrewserong commented:
> I don't get at all why we must render all inner blocks twice to fix and alignment issue.

Thanks for the ping @ellatrix! The only one that gets rendered twice is the current instance of the loop — the double rendering is to avoid flickering when switching between instances. Without it, when clicking between instances, there's a flash of white as the preview renders. If the current instance is rendered for "real" as well as there being a hidden preview version of it, then the preview can be shown / hidden instantaneously. Not the best approach, but resulted in a smoother UI. That could totally be revisited if there are better approaches now, though.

For a bit more content: and my memory is a little fuzzy on this particular PR since it's a couple of years old now, but my understanding is that the way that the block works is that it should only render the "real" inner blocks once, for the instance selected. And all other instances of the loop should render a preview instead. If we accidentally render each child as "real" inner blocks (as in #60535) then we wind up with a bunch of duplicate controls in the UI (in both the block toolbar and inspector controls):

<img width="1273" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/1aa69902-04c0-453d-8963-af655558daf4">

I believe using previews for the non-selected instances of the loop was a workaround for that problem. The previews pre-date this particular PR — this one was an attempt to work around some of the downsides of the preview approach prior to this PR (alignment display, reduce flickering etc).
